server/pxf-hdfs/src/main/java/org/greenplum/pxf/plugins/hdfs/utilities/AvroUtilities.java

```diff
@@ -0,0 +1,287 @@
+            try {
+                return searchForFile(file);
+            } catch (UnsupportedEncodingException e) {
+                e.printStackTrace();
+                return null;
+            }
+        };
+    }
+
+    private static AvroUtilities instance = new AvroUtilities();
+
+    // constructor for use in test
+    AvroUtilities(FileSearcher fileSearcher) {
+        this.fileSearcher = fileSearcher;
+    }
+
+    public static AvroUtilities getInstance() {
+        return instance;
+    }
+
+    /**
+     * All-purpose method for obtaining an Avro schema based on the request context and
+     * HCFS config.
+     *
+     * @param context
+     * @param configuration
+     * @return
+     */
+    public Schema obtainSchema(RequestContext context, Configuration configuration, HcfsType hcfsType) {
+        Schema schema = (Schema) context.getMetadata();
+
+        if (schema != null) {
+            return schema;
+        }
+        try {
+            schemaPath = context.getDataSource();
+            schema = readOrGenerateAvroSchema(context, configuration, hcfsType);
+        } catch (Exception e) {
+            throw new RuntimeException("Failed to obtain Avro schema from '" + schemaPath + "'", e);
+        }
+        context.setMetadata(schema);
+        return schema;
+    }
+
+    private Schema readOrGenerateAvroSchema(RequestContext context, Configuration configuration, HcfsType hcfsType) throws IOException, URISyntaxException {
+        // user-provided schema trumps everything
+        String userProvidedSchemaFile = context.getOption("SCHEMA");
+        if (userProvidedSchemaFile != null) {
+            schemaPath = userProvidedSchemaFile;
+            if (userProvidedSchemaFile.matches("^.*\\.avsc$")) {
+                return readSchemaFromUserProvidedJson(context, configuration, userProvidedSchemaFile, hcfsType);
+            }
+            return readSchemaFromUserProvidedAvroBinary(context, configuration, userProvidedSchemaFile);
+        }
+
+        // if we are writing we must generate the schema if there is none to read
+        if (context.getRequestType() == RequestContext.RequestType.WRITE_BRIDGE) {
+            return generateSchema(context.getTupleDescription());
+        }
+
+        // reading from external: get the schema from data source
+        return readSchemaFromAvroDataSource(configuration, context.getDataSource());
+    }
+
+    private Schema readSchemaFromUserProvidedJson(RequestContext context, Configuration configuration, String schemaName, HcfsType hcfsType) throws IOException, URISyntaxException {
+
+        InputStream schemaStream = null;
+        // try searching local disk (full path name and relative path in classpath) first.
+        // by searching locally first we can avoid an extra call to external (fs.exists())
+        try {
+            File file = fileSearcher.searchForFile(schemaName);
+            if (file == null) {
+                Path path = new Path(schemaName);
```

Comment:
You need to use HcfsType to get the URI string. I realize that there is no method in HcfsType, but it needs to be added there.
```suggestion
                String uri = hcfsType.getDataUri(configuration, schemaName);
```